[node-core-library] Moving @types/node to an optional peerDependency causes errors if the version doesn't match

<!-- Have a question?  Before creating an issue, ask in the chat room: https://rushstack.zulipchat.com/ -->

<!--------------------------------------------------------------------------
👉 STEP 1: Write a concise but specific issue title in the box above.
     Replace "the-package-name" with the project name without its NPM scope.
     For example, to report a problem for the "@rushstack/ts-command-line"
     project, your title might look like:

     "[ts-command-line] Using --example switch causes TypeError"
--------------------------------------------------------------------------->

## Summary
`@types/node` was moved to an optional peerDependency in https://github.com/microsoft/rushstack/pull/3937.  This caused `rush update --full` in our monorepo to fail with the following errors:

```
 ERR_PNPM_PEER_DEP_ISSUES  Unmet peer dependencies

├─┬ @rush-temp/abort-controller
│ └─┬ @microsoft/api-extractor
│   └─┬ @microsoft/api-extractor-model
│     └─┬ @rushstack/node-core-library
│       └── ✕ unmet peer @types/node@^12.20.24: found 14.18.36
└─┬ @rush-temp/schema-registry-avro
  └─┬ @microsoft/api-extractor
    └─┬ @microsoft/api-extractor-model
      └─┬ @rushstack/node-core-library
        └── ✕ unmet peer @types/node@^12.20.24: found 16.18.11
```

The peer dependency range was upgraded from `^12.20.24` to `^14.18.36` in https://github.com/microsoft/rushstack/pull/3928, but this doesn't change the core issue, since the packages are still incompatible with `@types/node@^16` and `@types/node@^18`.

`pnpm` does not issue warnings/errors for missing optional peer dependencies, but it does issue a warning/error for peer dependencies that a present but a non-matching version.  I'm not sure if this is by design or a bug in `pnpm`, and I've filed https://github.com/pnpm/pnpm/issues/6013 to find out.

If this is by design in `pnpm`, I think Rush should either:

1. Remove `@types/node` as an optional peerDependency
2. Expand the range of allowed versions, from `^14.18.36` to something like `^14.18.36 || ^16.18.11 || ^18.11.18`, to allow all the current Node LTS versions.

<!--------------------------------------------------------------------------
👉 STEP 2: In a few sentences, please explain:

     What were you trying to accomplish?
     What action did you perform that ran into trouble?
     What went wrong?
--------------------------------------------------------------------------->

## Repro steps

`pnpm add @rushstack/node-core-library@3.55.0 @types/node@16.18.11`

<!--------------------------------------------------------------------------
👉 STEP 3: If your issue is a feature request and not a bug, delete this
     "Repro steps" section and skip to STEP 6.

👉 STEP 4: In many cases we can investigate bugs much faster if you include:
     The URL for a simplified Git branch that reproduces the problem.
     Step by step instructions for how to build the branch and see the error.

👉 STEP 5: It's also helpful to include an "expected" and "actual" result.
     But if that's not relevant, feel free to delete those fields.
--------------------------------------------------------------------------->

  **Expected result:** <!-- What you expected these steps to accomplish -->
No warnings or errors.

  **Actual result:** <!-- If an error occurred, include the full error message text and any call stack. -->
```
dependencies:
+ @rushstack/node-core-library 3.55.0
+ @types/node 16.18.11 (18.11.18 is available)

 WARN  Issues with peer dependencies found
.
└─┬ @rushstack/node-core-library 3.55.0
  └── ✕ unmet peer @types/node@^14.18.36: found 16.18.11
```

## Details

<!--------------------------------------------------------------------------
👉 STEP 6: Provide any additional information you think might be helpful:

     What do you think is the cause of this problem?
     How do you think we should fix this?
--------------------------------------------------------------------------->

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| Package name: | @rushstack/node-core-library |
| Package version? | 3.55.0 |
| Operating system? | Linux |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | v18.13.0 |

--- a/common/changes/@rushstack/module-minifier/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/module-minifier/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/module-minifier"
+    }
+  ],
+  "packageName": "@rushstack/module-minifier",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/node-core-library/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/node-core-library/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/node-core-library"
+    }
+  ],
+  "packageName": "@rushstack/node-core-library",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/terminal/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/terminal/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/terminal"
+    }
+  ],
+  "packageName": "@rushstack/terminal",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/typings-generator/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/typings-generator/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/typings-generator"
+    }
+  ],
+  "packageName": "@rushstack/typings-generator",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/webpack4-localization-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/webpack4-localization-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/webpack4-localization-plugin"
+    }
+  ],
+  "packageName": "@rushstack/webpack4-localization-plugin",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/webpack4-module-minifier-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/webpack4-module-minifier-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/webpack4-module-minifier-plugin"
+    }
+  ],
+  "packageName": "@rushstack/webpack4-module-minifier-plugin",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/webpack5-localization-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/webpack5-localization-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/webpack5-localization-plugin"
+    }
+  ],
+  "packageName": "@rushstack/webpack5-localization-plugin",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/webpack5-module-minifier-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/webpack5-module-minifier-plugin/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/webpack5-module-minifier-plugin"
+    }
+  ],
+  "packageName": "@rushstack/webpack5-module-minifier-plugin",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/common/changes/@rushstack/worker-pool/node-peerdep-wildcard_2023-02-04-23-42.json
+++ b/common/changes/@rushstack/worker-pool/node-peerdep-wildcard_2023-02-04-23-42.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Change the peer dependency selector on `@types/node` to a wildcard (`*`).",
+      "type": "patch",
+      "packageName": "@rushstack/worker-pool"
+    }
+  ],
+  "packageName": "@rushstack/worker-pool",
+  "email": "iclanton@users.noreply.github.com"
+}

--- a/libraries/module-minifier/package.json
+++ b/libraries/module-minifier/package.json
@@ -30,7 +30,7 @@
     "@types/serialize-javascript": "5.0.2"
   },
   "peerDependencies": {
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/node": {

--- a/libraries/node-core-library/package.json
+++ b/libraries/node-core-library/package.json
@@ -36,7 +36,7 @@
     "@types/semver": "7.3.5"
   },
   "peerDependencies": {
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/node": {

--- a/libraries/terminal/package.json
+++ b/libraries/terminal/package.json
@@ -29,7 +29,7 @@
     "colors": "~1.2.1"
   },
   "peerDependencies": {
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/node": {

--- a/libraries/typings-generator/package.json
+++ b/libraries/typings-generator/package.json
@@ -32,7 +32,7 @@
     "@types/node": "14.18.36"
   },
   "peerDependencies": {
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/node": {

--- a/libraries/worker-pool/package.json
+++ b/libraries/worker-pool/package.json
@@ -23,7 +23,7 @@
     "@types/node": "14.18.36"
   },
   "peerDependencies": {
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/node": {

--- a/webpack/webpack4-localization-plugin/package.json
+++ b/webpack/webpack4-localization-plugin/package.json
@@ -18,7 +18,7 @@
     "@rushstack/set-webpack-public-path-plugin": "^3.3.89",
     "@types/webpack": "^4.39.0",
     "webpack": "^4.31.0",
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@rushstack/set-webpack-public-path-plugin": {

--- a/webpack/webpack4-module-minifier-plugin/package.json
+++ b/webpack/webpack4-module-minifier-plugin/package.json
@@ -23,7 +23,7 @@
     "@types/webpack-sources": "*",
     "webpack": "^4.31.0",
     "webpack-sources": "~1.4.3",
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "peerDependenciesMeta": {
     "@types/webpack": {

--- a/webpack/webpack5-localization-plugin/package.json
+++ b/webpack/webpack5-localization-plugin/package.json
@@ -17,7 +17,7 @@
   },
   "peerDependencies": {
     "webpack": "^5.68.0",
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "dependencies": {
     "@rushstack/localization-utilities": "workspace:*",

--- a/webpack/webpack5-module-minifier-plugin/package.json
+++ b/webpack/webpack5-module-minifier-plugin/package.json
@@ -21,7 +21,7 @@
   "peerDependencies": {
     "@rushstack/module-minifier": "*",
     "webpack": "^5.68.0",
-    "@types/node": "^14.18.36"
+    "@types/node": "*"
   },
   "dependencies": {
     "@rushstack/worker-pool": "workspace:*",

